src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp

```diff
@@ -0,0 +1,259 @@
+  dense_stepper.calc_state(t, quaternion_to_integrate);
+}
+
+template <size_t MaxDerivReturned>
+boost::math::quaternion<double>
+QuaternionFunctionOfTime<MaxDerivReturned>::setup_func(
+    double t) const noexcept {
+  // Populate missing quaternions
+  update_stored_info();
+
+  // Get quaternion and time at closest time before t
+  const auto& stored_info_at_t0 =
+      stored_info_from_upper_bound(t, stored_quaternions_and_times_);
+  std::array<DataVector, 1> quat_to_integrate =
+      stored_info_at_t0.stored_quantities;
+
+  // Solve the ode and store the result in quat_to_integrate
+  solve_quaternion_ode(t, stored_info_at_t0.time, quat_to_integrate);
+
+  // Transform DataVector quaternion to boost quaternion for easy manipulation
+  boost::math::quaternion<double> quat =
+      datavector_to_quaternion<double>(quat_to_integrate[0]);
+  normalize_quaternion(make_not_null(&quat));
+
+  return quat;
+}
+
+template <size_t MaxDerivReturned>
+std::array<DataVector, 1> QuaternionFunctionOfTime<MaxDerivReturned>::func(
+    double t) const noexcept {
+  boost::math::quaternion<double> quat = setup_func(t);
+
+  return std::array<DataVector, 1>{quaternion_to_datavector(quat)};
+}
+
+template <size_t MaxDerivReturned>
+std::array<DataVector, 2>
+QuaternionFunctionOfTime<MaxDerivReturned>::func_and_deriv(
+    double t) const noexcept {
+  boost::math::quaternion<double> quat = setup_func(t);
+
+  // Get omega and however many derivatives we need
+  std::array<DataVector, 1> omega_func = omega_f_of_t_ptr_->func(t);
+
+  boost::math::quaternion<double> omega =
+      datavector_to_quaternion<double>(omega_func[0]);
+
+  boost::math::quaternion<double> dtquat = 0.5 * quat * omega;
+
+  return std::array<DataVector, 2>{quaternion_to_datavector(quat),
+                                   quaternion_to_datavector(dtquat)};
+}
+
+template <size_t MaxDerivReturned>
+std::array<DataVector, 3>
+QuaternionFunctionOfTime<MaxDerivReturned>::func_and_2_derivs(
+    double t) const noexcept {
+  boost::math::quaternion<double> quat = setup_func(t);
+
+  // Get omega and however many derivatives we need
+  std::array<DataVector, 2> omega_func_and_deriv =
+      omega_f_of_t_ptr_->func_and_deriv(t);
+
+  boost::math::quaternion<double> omega =
+      datavector_to_quaternion<double>(omega_func_and_deriv[0]);
+  boost::math::quaternion<double> dtomega =
+      datavector_to_quaternion<double>(omega_func_and_deriv[1]);
+
+  boost::math::quaternion<double> dtquat = 0.5 * quat * omega;
+  boost::math::quaternion<double> dt2quat =
+      0.5 * (dtquat * omega + quat * dtomega);
+
+  return std::array<DataVector, 3>{quaternion_to_datavector(quat),
+                                   quaternion_to_datavector(dtquat),
+                                   quaternion_to_datavector(dt2quat)};
+}
+
+template <size_t MaxDerivReturned>
+std::array<DataVector, 4>
+QuaternionFunctionOfTime<MaxDerivReturned>::func_and_3_derivs(
+    double t) const noexcept {
```

Comment:
I think this function should `static_assert` that MaxDerivReturned >=3.  Similar for the other functions that return a fixed number of derivatives.